### PR TITLE
[Adhoc] Replace the link to the Ad-hoc forums in Networking options to a PPSSPP Github Wiki page link

### PR DESCRIPTION
**Pull request by ghost**
The wiki page got created by me and contains focused info about setting up multiplayer and a link to PPSSPP Ad-hoc forums as well if someone still needs it.
The page could also be edited by anyone who wants to.
Hope I did it correctly.

### CODE REVIEW

**Comment by hrydgard** [comment]
I changed "over" to "with" in the title - can you relink to https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-with-PPSSPP ?

I think this is a good idea. Alternatively I could create a page on the site, but as a "permanent" link built into the app, either works fine, can just link further.

**Comment by hrydgard** [comment]
We're gonna lose translations for this one since the text changed... Considering waiting until it's time to start merging 1.12 stuff.

**Comment by leoxxx** [comment]
I want to translatate it Chinese and add some pictures. How to do it?

**Comment by MojoJojoDojo** [comment]
> https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-with-PPSSPP
> I want to translatate the page into Chinese and add some pictures. How to do it?

You can create another page for the chinese language and link to it thats what other wiki pages I saw did.
Not sure about pictures but I will add pictures later with hyperlinking to them.

**Comment by leoxxx** [comment]
@MojoJojoDojo 

> otherwise you will need to run PPSSPP as administrator/sudo.

I think it's not necessary in windows at least.

**Comment by MojoJojoDojo** [comment]
Yeah added some pictures now thanks.

**Comment by leoxxx** [comment]
> Yeah added some pictures now thanks.
> Also administrator rights are needed for Linux.. for using ports above 1024. and maybe some Windows users too.
Port above 1024 is not necessary in all game.

File: UI/GameSettingsScreen.cpp [hunk 1]
```diff
@@ -1069,7 +1069,7 @@ void RecreateActivity() {
 }
 
 UI::EventReturn GameSettingsScreen::OnAdhocGuides(UI::EventParams &e) {
-	LaunchBrowser("https://forums.ppsspp.org/forumdisplay.php?fid=34");
+	LaunchBrowser("https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-over-PPSSPP");
```
Comment (ANR2ME) [inline UI/GameSettingsScreen.cpp]: The landing page link may need to go through translation too isn't? if we wanted to support multiple wiki pages with different language each?

Comment (MojoJojoDojo) [reply UI/GameSettingsScreen.cpp]: I guess so but I dont know where the translation stuff is.

Comment (hrydgard) [reply UI/GameSettingsScreen.cpp]: That's a good point - and probably a good idea! Should name it with a second argument then. Like:

`x->T("MultiplayerLink", "http://....")`

**Comment by leoxxx** [comment]
@MojoJojoDojo 
What's the meaning "work over"?
https://www.merriam-webster.com/dictionary/work over
Which one?

**Comment by MojoJojoDojo** [comment]
For some reason I like to use "over" instead of "with" lol they are similar.

**Comment by leoxxx** [comment]
> For some reason I like to use "over" instead of "with" lol they are similar.
> Or use "via" .

I see.

**Comment by leoxxx** [comment]
@MojoJojoDojo 

> Forced First Connect (forcing blocking behavior on the first attempt to connect to get connected faster)

Can't understand this.
Have some more detail explanation?

I think it is blocking the the first attempt to connect behavior,not blocking first connect. Is it right?

**Comment by MojoJojoDojo** [comment]
@ANR2ME can explain more on that.
TBH it may be gone soon anyway as an option.

**Comment by ANR2ME** [comment]
Socket have 2 modes, blocking and non-blocking, with FFC if the game try to connect using non-blocking socket it will be forced to use blocking mode to make sure the socket to get connected before returning from HLE.

**Comment by leoxxx** [comment]
> non-blocking socket

Thanks. I see.

**Comment by leoxxx** [comment]
> Socket have 2 modes, blocking and non-blocking, with FFC if the game try to PtpConnect using non-blocking socket it will be forced to use blocking mode on the 1st attempt to PtpConnect, to make sure the socket to get connected before returning from PtpConnect HLE.
> 
> There are games (ie. GVG Next Plus) that will try to connect again after waiting for more than 5-10 seconds if the first attempt to connect failed (in the case the other side is not listening yet), and waiting for 5+ seconds will probably feels too long for most people, so FFC can prevent this waiting by making sure the first attempt to be successful (but the other side need to be ready to accept the connection within 200 - 500ms since it will only be blocked for that duration)
> 
> But after the fix to The Warriors, we get an early `connect` during PtpOpen so it increases the chance for PtpConnect to be successful at the time the game using it, thus FFC may no longer be needed, but not sure whether FFC will still be needed for GvG over high latency internet or not tho.
> 
> PS: i'm planning to remove this workaround if someone can confirm that FFC makes no difference playing GvG over the internet, since i can't test any game over the internet my self.

Thanks. I see.

**Comment by leoxxx** [comment]
@MojoJojoDojo 
I‘ve finished the page.
Try to change Port Offset value, but can't trigger the "Failed to Bind Port xxx". It seems not to affect the PPSSPP used port.
QQ截图20210213235510
@ANR2ME 
How to trigger it?

**Comment by ANR2ME** [comment]
> 
> 
> @MojoJojoDojo
> I‘ve finished the page.
> Try to change Port Offset value, but can't trigger the "Failed to Bind Port xxx". It seems not to affect the PPSSPP used port.
> QQ截图20210213235510
> @ANR2ME
> How to trigger it?

PPSSPP is reusing the port, so i can bind some ports that also set to be reused, the only way to trigger it is to use a port that was exclusively used for one socket only or if it's not allowed by the OS (ie. ports 1024 or lower on Android)

**Comment by leoxxx** [comment]
> > @MojoJojoDojo
> > I‘ve finished the page.
> > Try to change Port Offset value, but can't trigger the "Failed to Bind Port xxx". It seems not to affect the PPSSPP used port.
> > QQ截图20210213235510
> > @ANR2ME
> > How to trigger it?
> 
> PPSSPP is reusing the port, so it can bind some ports that already used but also set to be reused, the only way to trigger it is to use a port that was exclusively used for one socket only or if it's not allowed by the OS (ie. ports 1024 or lower on Android, or linux without sudo)
> 
> For example if you runs games that use low ports such as GTA VCS or DBZ Shinbudokai 2 with port offset < 900 on Android it will trigger the error message.

THX.
I'll try it.

**Comment by leoxxx** [comment]
@MojoJojoDojo 
https://github.com/hrydgard/ppsspp/issues/14121
All Chinese pictures have be added in this issues.
If you want to add more Engilsh pictures, I can do it. But I can't add the UPnP and DMZ for router's pictures, because I haven't English router.

File: UI/GameSettingsScreen.cpp (hunk 1 above)
Comment (leoxxx) [reply UI/GameSettingsScreen.cpp]: > if we wanted to use multiple wiki pages with different language each?
I think so.

**Comment by MojoJojoDojo** [comment]
Yeah you can add them @leoxxx

**Comment by leoxxx** [comment]
@MojoJojoDojo 
I'll do it later.
BTW. I can't receive the notifications someone at me.
Here is my settings.
QQ截图20210214183641

**Comment by leoxxx** [comment]
@MojoJojoDojo 
I've finished them.

**Comment by hrydgard** [comment]
Please make it use i18n so we can have different links for different languages, using the ini files. Other than that, I'm happy to merge this now.

**Comment by leoxxx** [comment]
> Please make it use i18n so we can have different links for different languages, using the ini files. Other than that, I'm happy to merge this now.

I just know Chinese , a little English and a little Japanese.
If somebodys can translate the English or Chinese one to other languages' ones, I'll try to reference English or Chinese one to add more pictures.

**Comment by MojoJojoDojo** [comment]
> Please make it use i18n so we can have different links for different languages, using the ini files. Other than that, I'm happy to merge this now.

Sorry only noticed this now.
How should I do this change?

File: UI/GameSettingsScreen.cpp (hunk 1 above)
Comment (MojoJojoDojo) [reply UI/GameSettingsScreen.cpp]: So should I do it LaunchBrowser(x->T("MultiplayerLink","https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-over-PPSSPP")); ?

**Comment by hrydgard** [comment]
It is important to some users, and I want to make more of the website and stuff internationalization-friendly in the future. It's also very easy to do:

```
auto n = GetI18NCategory("Networking");
LaunchBrowser(n->T("MultiplayerHowToURL", "https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-with-PPSSPP"));
```

(with reservation for any string type errors).

**Comment by ANR2ME** [comment]
Btw why one of the screenshot in multiplayer wiki have this weird looking text?
https://user-images.githubusercontent.com/5364709/107880943-d8ae4880-6f1c-11eb-8a8f-46e45a533c20.png

**Comment by unknownbrackets** [comment]
The translation files allow you to configure a different font, instead of Roboto.  That screenshot is using such an alternate font.

-[Unknown]

**Comment by leoxxx** [comment]
> Btw why one of the screenshot in multiplayer wiki have this weird looking text? reading this hurts my eyes LOL
> image

This screenshot is for step 1,2&3.2. So I add some texts.
Like this one for step 3.1
 image
Then how to modify it? Change the text size or color?
Do you say the PPSSPP texts?
 image
This is Chinese one.
I change the language to English, it auto change the font.

**Comment by MojoJojoDojo** [comment]
I dont want this merged yet because its missing the portable hotspot method (mentioned in this issue here : https://github.com/hrydgard/ppsspp/issues/13163)
But because I dont know how to use that method because I never used it.
So if someone knows how he could edit the wiki and I will fix the PR later.

**Comment by ANR2ME** [comment]
I don't think it's an issue anymore, since he was assuming that hotspot IP on Android is always 192.168.43.1 but apparently not (at least on Oneplus), and we didn't mention that and IP is always 192.168.43.1 in our wiki anyway. 
Players just need to use the IP detected by PPSSPP from the IP list.
And the issue he had with Windows was probably related to the wrong port being detected on Windows on MGS:PW which got fixed on  https://github.com/hrydgard/ppsspp/pull/14140

**Comment by MojoJojoDojo** [comment]
Yeah but I think this is the method people use when they have no access to Wi-fi no?
Otherwise how they connect their Android to PC if they only got 3G/4G/LTE on both of their devices ?

**Comment by ANR2ME** [comment]
The IP 192.168.x.x should be available on the IP List to be selected on Android when that android device is acting as a hotspot, so the player who act as hostpot only need to select the IP, instead of assuming it will always be 192.168.43.1, and tells other players what is his hotspot IP (which is the IP being selected)

**Comment by MojoJojoDojo** [comment]
> It is important to some users, and I want to make more of the website and stuff internationalization-friendly in the future. It's also very easy to do:
> 
> ```
> auto n = GetI18NCategory("Networking");
> LaunchBrowser(n->T("MultiplayerHowToURL", "https://github.com/hrydgard/ppsspp/wiki/How-to-play-multiplayer-games-with-PPSSPP"));
> ```
> 
> (with reservation for any string type errors).

Sorry it took so much time lol but I think its ready now.
I wanted to explain how the portable hotspot method works but I dont use Android much so maybe someone else could do that.

**Comment by MojoJojoDojo** [comment]
@hrydgard Am I missing something?
I think I did what you said.
It's not merged yet.
Or maybe @ANR2ME could merge cause he got merging prviliages and it seems like a small change lol.

**Comment by ANR2ME** [comment]
The code looks alright, while the commit's naming aren't informative enough tho :) (ie. "Update GameSettingsScreen.cpp" while a lot of things could exist on GameSettingsScreen.cpp)

Also, you have tested that it worked properly and opening the correct page, right?

**Comment by MojoJojoDojo** [comment]
> The code looks alright, although commit's naming aren't informative enough tho :) (ie. "Update GameSettingsScreen.cpp" while a lot of things could exist on GameSettingsScreen.cpp)
> 
> Also, you have tested that it worked properly and opening the correct page, right?

Yes it works I tested it @ANR2ME .

**Comment by ANR2ME** [comment]
Okay, i'll merge this, since this is just a small changes.

**Comment by leoxxx** [comment]
@hrydgard
It's not goto https://github.com/hrydgard/ppsspp/wiki/如何使用PPSSPP多人联机游戏  the Chinese page.

**Pull request merged**